BlueDBOnDisk/src/main/java/org/bluedb/disk/collection/index/extractors/OverlappingTimeSegmentsKeyExtractor.java

```diff
@@ -0,0 +1,41 @@
+package org.bluedb.disk.collection.index.extractors;
+
+import java.io.Serializable;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import org.bluedb.api.keys.BlueKey;
+import org.bluedb.api.keys.LongTimeKey;
+import org.bluedb.disk.StreamUtils;
+import org.bluedb.disk.segment.ReadableSegmentManager;
+
+/**
+ * Returns the start grouping number of all segments that the record overlaps with the exception of the first segment. 
+ * Given a timeframe, you should be able to use this index to easily find all of the segments that contain records that
+ * started before it but overlap into it.
+ */
+public class OverlappingTimeSegmentsKeyExtractor<V extends Serializable> implements DefaultTimeKeyExtractor<LongTimeKey, V> {
+
+	private static final long serialVersionUID = 1L;
+
+	@Override
+	public Class<LongTimeKey> getType() {
+		return LongTimeKey.class;
+	}
+	
+	@Override
+	public List<LongTimeKey> extractKeys(BlueKey key, ReadableSegmentManager<V> segmentManager) {
+		LinkedList<LongTimeKey> keys = StreamUtils.stream(segmentManager.getPathManager().getAllPossibleSegmentStartGroupingNumbers(key))
+			.map(LongTimeKey::new)
+			.collect(Collectors.toCollection(LinkedList::new));
+		keys.remove(0);

```

Comment:
Why are we removing the `0` index here? Does that need to happen every time. Is it possible that the stream gives us an empty list?

Reply:
All keys have one grouping number. The method it calls will always at least return that grouping number. If the key overlaps more than one segment then it would also return grouping numbers for those segments. We are removing the first grouping number since we don't care about tracking that in the index. We know how to find a value in the segment that it starts in. We want to know what segments it overlaps, so that when we're querying those segments we can easily find the records that started sometime in the past and overlap into our segment.

That being said, it is irrelevant now. I just pushed a commit that refactors that a little bit and does it in a different way. Sorry about that, I didn't realize I forgot to push that last commit.

Reply:
Ha, well I will look at the new commit. Thanks for the explanation either way